### PR TITLE
HDDS-2321. Ozone Block Token verify should not apply to all datanode …

### PR DESCRIPTION
**Pull request by xiaoyuyao**
## What changes were proposed in this pull request?

* Change the TokenVerifier interface to check the command type and the block id.
* Token verification based on token encode in the command done inside HddsDispatcher. 
*  Remove the Grpc Client/Server CredentialInterceptor as it cannot fit into Ratis commands. 
* Added more unit test coverage on the Tokenverifier.
 * 

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-2321

## How was this patch tested?

Added Unit test testBlockTokenVerifier()
Update Unit test in TestSecureContainerServer.java 
ozone secure smoke test.

### CODE REVIEW

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/HddsUtils.java [hunk 1]
```diff
@@ -390,6 +391,52 @@ public static boolean isReadOnly(
...
+  }
+
+  /**
+   * Return the block ID of container commands that are related to blocks.
+   * @param msg container command
+   * @return block ID.
+   */
+  public static BlockID getBlockID(ContainerProtos.ContainerCommandRequestProto msg) {
+    switch (msg.getCmdType()) {
+    case ReadChunk:
```
[the remark this answers is not in the document]
Comment (xiaoyuyao) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/HddsUtils.java]: Good point. Fixed.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/security/token/BlockTokenVerifier.java [hunk 2]
```diff
@@ -55,68 +57,68 @@ private boolean isExpired(long expiryDate) {
...
-      // defer access mode, bcsid and maxLength check to container dispatcher
-      UserGroupInformation ugi = tokenId.getUser();
-      ugi.addToken(token);
-      ugi.setAuthenticationMethod(UserGroupInformation
-          .AuthenticationMethod.TOKEN);
-      return ugi;
-    } else {
-      return UserGroupInformation.createRemoteUser(user);
+    UserGroupInformation tokenUser = tokenId.getUser();
+    X509Certificate singerCert;
```
[the remark this answers is not in the document]
Comment (xiaoyuyao) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/security/token/BlockTokenVerifier.java]: Good catch. Fixed.

**Comment by xiaoyuyao** [comment]
Thanks @adoroszlai  and @anuengineer  for the review. The unit test and checkstyle issues are fixed.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/HddsUtils.java [hunk 3]
```diff
@@ -390,6 +391,52 @@ public static boolean isReadOnly(
...
+   * false if block token does not apply to the command.
+   *
+   */
+  public static boolean requireOmBlockToken(
+      ContainerProtos.Type cmdType) {
+    switch (cmdType) {
+    case ReadChunk:
+    case GetBlock:
+    case WriteChunk:
+    case PutBlock:
```
[the remark this answers is not in the document]
Comment (xiaoyuyao) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/HddsUtils.java]: I think the token for PutSmallFile/GetSmallFile are only used by Quadra but can potentially be used by OM to optimize certain workload. I've updated the PR as suggested.

**Comment by adoroszlai** [comment]
Thanks @xiaoyuyao for the update.

Acceptance test failure is unrelated, it happens on master, too.  Details: https://github.com/apache/hadoop-ozone/pull/11#issuecomment-549783637

Integration test failure is also unrelated, already tracked in Jira ([HDDS-2367](https://issues.apache.org/jira/browse/HDDS-2367) and [HDDS-2392](https://issues.apache.org/jira/browse/HDDS-2392)).

**Comment by xiaoyuyao** [comment]
Thanks @adoroszlai  for the confirmation. I will merge the PR shortly.

**Pull request closed without merging**